rs/ledger_suite/icrc1/ledger/src/lib.rs

```diff
@@ -503,36 +545,80 @@ impl<Tokens: TokensType> Ledger<Tokens> {
 
         ledger
     }
+
+    pub fn is_ready(&self) -> bool {
+        matches!(self.state, LedgerState::Ready)
+    }
+
+    pub fn migrate_one_allowance(&mut self) -> bool {
+        match self.approvals.allowances_data.pop_first_allowance() {
+            Some((account_spender, allowance)) => {
+                self.stable_approvals
+                    .allowances_data
+                    .set_allowance(account_spender, allowance);
+                true
+            }
+            None => false,
+        }
+    }
+
+    pub fn migrate_one_expiration(&mut self) -> bool {
+        match self.approvals.allowances_data.pop_first_expiry() {
+            Some((timestamp, account_spender)) => {
+                self.stable_approvals
+                    .allowances_data
+                    .insert_expiry(timestamp, account_spender);
+                true
+            }
+            None => false,
+        }
+    }
+
+    pub fn clear_arrivals(&mut self) {
+        self.approvals.allowances_data.clear_arrivals();
+    }
 }
 
-impl<Tokens: TokensType> LedgerContext for Ledger<Tokens> {
+impl LedgerContext for Ledger {
     type AccountId = Account;
-    type AllowancesData = HeapAllowancesData<Self::AccountId, Tokens>;
+    type AllowancesData = StableAllowancesData;
     type BalancesStore = BTreeMap<Self::AccountId, Tokens>;
     type Tokens = Tokens;
 
     fn balances(&self) -> &Balances<Self::BalancesStore> {
+        if !self.is_ready() {
+            ic_cdk::trap("The Ledger is not ready");
+        }
```

Comment:
nit: Would it make sense to have a `self.panic_if_not_ready()` function since the end result is the same regardless of the method called?

Reply:
Done!